Streaming and on_llm_new_token calling for LlamaCpp used for Alpaca

lukestanley opened the pull request:
## Summary

This adds streaming support to the LlamaCpp LLM class. It also improves the LlamaCpp class with a new `_get_parameters` method and updated tests to cover streaming tokens and the invocation of the `on_llm_new_token` callback.

## Changes

### langchain/llms/llamacpp.py


### tests/integration_tests/llms/test_llamacpp.py

- Added `test_llamacpp_streaming` to test streaming tokens from LlamaCpp
- Added `test_llamacpp_streaming_callback` to test that streaming correctly invokes the `on_llm_new_token` callback

I'm new so if I've missed something, please kindly point it out.

lukestanley committed:
Adds test for on_new_tokens callback use and stream generator

lukestanley committed:
Improve LlamaCpp with streaming support and new _prepare_params method. Update tests to cover streaming tokens, and add a test to verify correct invocation of on_llm_new_token callback.

lukestanley committed:
Adds async methods to llamacpp.py and stream generator method

lukestanley commented:
I am fixing some linting errors on this.

vowelparrot commented on langchain/llms/llamacpp.py:
```diff
@@ -195,3 +213,99 @@ def _call(self, prompt: str, stop: Optional[List[str]] = None) -> str:
+            for generation in self.client(prompt=prompts[0], stream=stream, **params):
+                chunk = generation['choices'][0]['text']
+            if self.verbose:
+                print(chunk, end="", flush=True)
```
Please remove the `print` statements

vowelparrot commented on langchain/llms/llamacpp.py:
```diff
@@ -195,3 +213,99 @@ def _call(self, prompt: str, stop: Optional[List[str]] = None) -> str:
+                chunk = generation['choices'][0]['text']
+            if self.verbose:
+                print(chunk, end="", flush=True)
+            yield chunk
```
This `yield` statement ought to be indented once more, right?

vowelparrot commented on tests/integration_tests/llms/test_llamacpp.py:
```diff
@@ -32,3 +35,32 @@ def test_llamacpp_inference() -> None:
+    assert isinstance(generator, Generator)
+
+    for token in generator:
+        assert isinstance(token, str)
```
We should collect the tokens and assert the result is non-empty

vowelparrot reviewed (changes requested):
Could you add a streaming examplesto the notebook as well? Thank you for the PR!

lukestanley commented:
Thank you for looking, I think I made a mistake when fixing the merge conflict! I will fix mypy issues and improve the tests. I'm learning more about the way the LLM classes are meant to work! I am converting this back to draft!

lukestanley committed:
Removes async methods, rewrites parameter helper and uses stream method with plain generator (not async), and improves tests to check length, and assert same format as OpenAI LLM class

lukestanley commented:
> Could you add a streaming examplesto the notebook as well? Thank you for the PR!

I can try doing this, also I think a gif could be helpful with some chain being used.

lukestanley commented:
Thanks for earlier reviews @vowelparrot  this should run much more happily now. I've simplified it a fair bit by throwing out the async stuff for now and just basing it around the stream generator. It also is more similar to the OpenAI LLM class now. I've implemented length checks as suggested too. Linting is passing for me as is the integration test.

vowelparrot commented on langchain/llms/llamacpp.py:
```diff
@@ -167,31 +195,35 @@ def _call(self, prompt: str, stop: Optional[List[str]] = None) -> str:
-            params["stop_sequences"] = self.stop
-        elif stop:
-            params["stop_sequences"] = stop
+        print("Prompt: ", prompt)
```
Will want to remove print statements before landing

lukestanley replied:
Oops, I was testing with load_summarize_chain and forgot this was in! Fixed now.

vowelparrot commented on langchain/llms/llamacpp.py:
```diff
@@ -154,6 +157,31 @@ def _llm_type(self) -> str:
+        
+        params = self._default_params
+        
+        # llamaa_cpp expects the "stop" key not this, so we remove it:
```
```suggestion
         # llama_cpp expects the "stop" key not this, so we remove it:
```

vowelparrot reviewed (commented):
Streaming is working locally - well done! Will have to clean up some of the linting errors but looks mostly there

lukestanley committed:
Fix typo, removes print

lukestanley committed:
Linting

lukestanley committed:
Documents new LlamaCpp stream method

lukestanley committed:
Stream by default for LlamaCpp

lukestanley committed:
Adds token callback demo for LlamaCpp

lukestanley commented:
> Could you add a streaming examplesto the notebook as well? Thank you for the PR!

I improved the docstring for the stream method, which is not yet demoed but is very similar except it can include the logprobs. 

I added the callback demo. If people play it themselves they should see something like this (starts at 14 seconds):

https://user-images.githubusercontent.com/306671/233401708-be27b8a3-08d0-4361-a198-d8aa381e38fd.mp4

@vowelparrot

lukestanley committed:
Fix typo

lukestanley committed:
Linting and typo fix

vowelparrot reviewed (approved):
Beautiful. Thank you for your patience!
